Firestore/core/test/firebase/firestore/testutil/async_testing.cc

```diff
@@ -36,6 +40,62 @@ std::shared_ptr<util::AsyncQueue> AsyncQueueForTesting() {
   return std::make_shared<AsyncQueue>(ExecutorForTesting("worker"));
 }
 
+// MARK: - Expectation
+
+Expectation::Expectation() : promise_(std::make_shared<std::promise<void>>()) {
+}
+
+void Expectation::Fulfill() {
+  promise_->set_value();
+}
+
+std::function<void()> Expectation::AsCallback() const {
+  // Additional copy required because putting `promise_` in the capture list
+  // is not allowed. Generalized capture in C++14 would allow direct
+  // initialization.
+  auto promise = promise_;
+  return [promise] { promise->set_value(); };
+}
+
+std::future<void> Expectation::get_future() const {
```

Comment:
This is a destructive operation, so it probably shouldn't be `const` (even though the language rules allow modifying a value through a pointer inside a const function). `std::promise::get_future` isn't `const`.

Reply:
I've changed this around to just use `std::shared_future` instead of `std::future` on the grounds that efficiency doesn't matter here and the move-only nature of `std::future` harms the design considerably. With `std::shared_future` this becomes just an accessor declared in the header and no longer has any consequences as far as reuse goes.